### ISSUE TITLE
post-processing

### ISSUE THREAD

**Issue by zhao181** [HRNet/HRNet-Facial-Landmark-Detection#26]
Dear author, how can I understand the post-processing part in the decode_preds function？

    # pose-processing
    for n in range(coords.size(0)):
        for p in range(coords.size(1)):
            hm = output[n][p]
            px = int(math.floor(coords[n][p][0]))
            py = int(math.floor(coords[n][p][1]))
            if (px > 1) and (px < res[0]) and (py > 1) and (py < res[1]):
                diff = torch.Tensor([hm[py - 1][px] - hm[py - 1][px - 2], hm[py][px - 1]-hm[py - 2][px - 1]])
                coords[n][p] += diff.sign() * .25
    coords += 0.5
    preds = coords.clone()

**Comment by PlumedSerpent** [comment]
@zhao181 Hi, I can try to explain it, but  I' m not the author, so maybe I'll make some mistakes in some 
details.
As you know, our input image is 256x256 pixel, while the sizes of output heatmaps are 64x64, which is only a quarter of the original image. Then it comes to a problem when you transform your output to your predicted coordinates.
Let's take an example, the ground-truth coordinate of a certain landmark is (242,242), then we transform it into a heatmap(64x64), we just divide it by 4, so it turns to (60.5,60.5). However, the index of an image variable must be a int , so we just take it as (60,60). Suppose our network can predict heatmaps very accurately, then the heatmap will be a gaussian heatmap with its center on (60,60), then if you just multiply it by 4, you can only get (240,240), which will have 2 pixels error on each direction.
How can we solve this problem? First, it takes some tricks when generating heatmap ground-truth, you can look into the code in lib/dataset/* and lib/util/transform.py. Second, you can make the predicted heatmap output some float number. In your above code, hm is a predicted heatmap of one landmark of one sample, "coords" is the sorted index of the heatmap, so you can see px and py is just the output we need( if you don't care about the error I said). Next, the author checks if the px and py are not on the boundaries of the heatmap( in case it throws errors in the following operation).  Then the author checks the symmetry of the heatmap. For example, if the left side hm[py - 1][px-2] is larger than the right side hm[py - 1][px], then maybe we should move the center to left a little ( the reason why we take the py-1-th row is in py-th row the lefe side is surely smaller than the right side, and it has no help to the finetune of coords),  otherwise, we can see that now the px is a pretty good center(because no matter we choose which row, the px-th pixel is still larger), then we move to right a little, why? because we transform (240/241/242）into 64, you can see in most cases (i.e. 241/242) we better turn 64 into 64.25, even though 64 is exactly the ground-truth. 
As for the line "coords += 0.5", I  believe it's related to the ground-truth generating processing but  i'm not sure. @wondervictor Would you please to explain it?

**Comment by oftenliu** [comment]
do you know it now?

**Comment by zhao181** [comment]
No I don't know

**Comment by jiangtaoo2333** [comment]
2020.4.23. do you understand now.......

**Comment by lzcomeon** [comment]
Anyone can help?

**Comment by kuldeepbrd1** [comment]
It is a common practice to compensate for the quantization effect. Since the landmark is encoded in a heatmap of lower resolution (generally 1/4th) than the original image, when recovering the coordinates after inference, a sub-pixel shift towards the next highest pixel value is used. 

From [Stacked Hourglass Networks for Human Pose Estimation](https://arxiv.org/pdf/1603.06937.pdf), pg 8:
>  To improve performance at high precision thresholds the prediction is offset by a quarter of a pixel in the direction of its next highest neighbor before transforming back to the original coordinate space of the image.

Some details are also provided in [Distribution-Aware Coordinate Representation for Human Pose Estimation](https://arxiv.org/pdf/1910.06278v1.pdf) in **Coordinate Decoding** section.

Hope this helps!

**Comment by jiangtaoo2333** [comment]
@kuldeepbrd1  really thanks for your help.

**Comment by lzcomeon** [comment]
> It is a common practice to compensate for the quantization effect. Since the landmark is encoded in a heatmap of lower resolution (generally 1/4th) than the original image, when recovering the coordinates after inference, a sub-pixel shift towards the next highest pixel value is used.
> 
> From [Stacked Hourglass Networks for Human Pose Estimation](https://arxiv.org/pdf/1603.06937.pdf), pg 8:
> 
> > To improve performance at high precision thresholds the prediction is offset by a quarter of a pixel in the direction of its next highest neighbor before transforming back to the original coordinate space of the image.
> 
> Some details are also provided in [Distribution-Aware Coordinate Representation for Human Pose Estimation](https://arxiv.org/pdf/1910.06278v1.pdf) in **Coordinate Decoding** section.
> 
> Hope this helps!

Thanks a lot

**Comment by MengHao666** [comment]
> It is a common practice to compensate for the quantization effect. Since the landmark is encoded in a heatmap of lower resolution (generally 1/4th) than the original image, when recovering the coordinates after inference, a sub-pixel shift towards the next highest pixel value is used.
> 
> From [Stacked Hourglass Networks for Human Pose Estimation](https://arxiv.org/pdf/1603.06937.pdf), pg 8:
> 
> > To improve performance at high precision thresholds the prediction is offset by a quarter of a pixel in the direction of its next highest neighbor before transforming back to the original coordinate space of the image.
> 
> Some details are also provided in [Distribution-Aware Coordinate Representation for Human Pose Estimation](https://arxiv.org/pdf/1910.06278v1.pdf) in **Coordinate Decoding** section.
> 
> Hope this helps!

Hi，thanks for the reference. Why do we need to shift 0.5 pixel in[ here](https://github.com/HRNet/HRNet-Facial-Landmark-Detection/blob/f776dbe8eb6fec831774a47209dae5547ae2cda5/lib/core/evaluation.py#L77) ?

**Comment by kuldeepbrd1** [comment]
> Hi，thanks for the reference. Why do we need to shift 0.5 pixel in[ here](https://github.com/HRNet/HRNet-Facial-Landmark-Detection/blob/f776dbe8eb6fec831774a47209dae5547ae2cda5/lib/core/evaluation.py#L77) ?

@MengHao666 The 1/4 offset is made from the center of the heatmap pixel. 
The center of the heatmap pixel with coordinates (x,y) is 0.5+(x,y).

See the image below with green point being the final pred with sub-pixel shift. 
[I made a slight misrepresentation, the sub-pixel shift should be- (x' +/- 0.25, y' +/- 0.25)] as it can be in any direction]

Someone can maybe confirm this just to be sure.
![image](https://user-images.githubusercontent.com/31183061/165323560-2dd0970b-18bc-4268-9dd6-14166eaba049.png)

**Comment by MengHao666** [comment]
> any directio

In my own understanding, the `+0.5` is for the interpolation to decrease the final scaled  error. The 1/4 offeset is considering the heatmap , but do you thnk the `+0.5` and `0.25` are dpendent on experience, that is something hyper-parameters？

**Comment by MengHao666** [comment]
> any directio

@kuldeepbrd1  In my own understanding, the `+0.5` is for the interpolation to decrease the final scaled  error. The 1/4 offeset is considering the heatmap , but do you thnk the `+0.5` and `0.25` are dpendent on experience, that is something hyper-parameters？

**Comment by MengHao666** [comment]
Training on my own dataset, I have a very interesting result here with the same so-called post-processing . In the evaluation process, the `NME' first drop and then increase at the very begining 5th epoch. However, the landmark heatmap loss always decrease. I couldn't figure it out. Why? Anyone have any idea?
![image](https://user-images.githubusercontent.com/47114466/166615490-e70b1fda-a327-455a-8cfd-6c232e38a814.png)
